core/src/replay_stage.rs

```diff
@@ -3074,6 +3074,37 @@ impl ReplayStage {
 
                                 tower.vote_state.root_slot = bank_vote_state.root_slot;
                                 tower.vote_state.votes = bank_vote_state.votes;
+
+                                let last_voted_slot = tower.vote_state.last_voted_slot().unwrap_or(
+                                    // If our local root is higher than the highest slot in `bank_vote_state` due to
+                                    // supermajority roots, then it's expected that the vote state will be empty.
+                                    // In this case we use the root as our last vote. This root cannot be None, because
+                                    // `tower.vote_state.last_voted_slot()` is None only if `tower.vote_state.root_slot`
+                                    // is Some.
```

Comment:
where is this logic guaranteed?

Reply:
`None <= None`. If `tower.vote_state.last_voted_slot()` is `None`, this means `bank_vote_state.last_voted_slot()` must have originally been `Some` in order to pass this check:
```
if bank_vote_state.last_voted_slot()
                                > tower.vote_state.last_voted_slot()
```
That means it must have been adjusted here:
```
bank_vote_state .votes .retain(|lockout| lockout.slot() > local_root);
```
For it to have been adjusted, it must have passed this check:
```
if let Some(local_root) = tower.vote_state.root_slot {
```
So it follows that iff `tower.vote_state.root_slot` is `Some` then `tower.vote_state.last_voted_slot()` is `None`

Reply:
ah yeah, i missed the logic where we adjust the root and filter out slots less than the root